src/compute/src/sink/persist_sink.rs

```diff
@@ -720,11 +731,38 @@ where
                                     persist_frontier
                                 );
                             }
-                            correction.with_correction_buffer(sink_metrics, sink_worker_metrics, |buffer| buffer.append(&mut data));
+
+                            if let Some(stash) = &mut desired_stash {
+                                for (d, t, r) in data {
+                                    stash.entry(t).or_insert_with(Vec::new).push((d, r));
+                                }
+                            } else {
+                                correction.with_correction_buffer(
+                                    sink_metrics,
+                                    sink_worker_metrics,
+                                    |buffer| buffer.append(&mut data),
+                                );
+                            }
 
                             continue;
                         }
                         Event::Progress(frontier) => {
+                            // Extract desired rows as positive contributions to `correction`.
+                            if let Some(stash) = &mut desired_stash {
+                                stash.retain(|time, data| {
```

Comment:
It seems a bit wasteful to revisit all timestamps every time the frontier advances, even though:
- for the case that inspired this PR (IIUC) this may be quite a lot of distinct timestamps;
- the btreemap is stored in timestamp order, so in theory it should be efficient to peel off just the smallest timestamps.

I think if our timestamps were partially ordered we might just have to eat this cost; but I think here we're working with our concrete totally-ordered timestamp type and we can just pop off the entries in order. Does that scan?

Reply:
We indeed only sink totally-ordered times to persist. However, AFAICT, the whole `persist_sink` is currently written to accept partially-ordered times, and I'd rather not change that assumption as part of this PR, mostly because I think it might be controversial.

Definitely something to keep it mind as a potential follow-up though!

Reply:
Fair enough! (Given that all the operators in this file hardcode to `mz_repr::Timestamp`, it's unclear to me to what extent this is true in practice... but also I can imagine it being controversial anyways, and it's certainly not worth holding up a PR over. 😅)

Reply:
I added this optimization now after all, because you are right that we use `mz_repr::Timestamp` everywhere already and it was easy to add.